File: server-mod/src/GiveUserPresetSptCommand.ts
import {inject, injectable} from "tsyringe";
import {ISptCommand} from "@spt/helpers/Dialogue/Commando/SptCommands/ISptCommand";
import {ItemHelper} from "@spt/helpers/ItemHelper";
import {ISendMessageRequest} from "@spt/models/eft/dialog/ISendMessageRequest";
import {ISptProfile, IUserDialogInfo} from "@spt/models/eft/profile/ISptProfile";
import {MailSendService} from "@spt/services/MailSendService";
import {ICloner} from "@spt/utils/cloners/ICloner";
import {SaveServer} from '@spt/servers/SaveServer';


@injectable()
export class GiveUserPresetSptCommand implements ISptCommand {
    private static commandRegex = /^spt give-user-preset ((([a-z]{2,5}) )?"(.+)"|\w+)$/;

    public constructor(
        @inject("ItemHelper") protected itemHelper: ItemHelper,
        @inject("MailSendService") protected mailSendService: MailSendService,
        @inject("PrimaryCloner") protected cloner: ICloner,
        @inject("SaveServer") protected saveServer: SaveServer,
    ) {
    }

    public getCommand(): string {
        return "give-user-preset";
    }

    public getCommandHelp(): string {
        return 'spt give-user-preset\n========\nSends items to the player through the message system.\n\n\tspt give-user-preset [weaponBuilds.Id]';
    }

    public performAction(commandHandler: IUserDialogInfo, sessionId: string, request: ISendMessageRequest): string {
        if (!GiveUserPresetSptCommand.commandRegex.test(request.text)) {
            this.mailSendService.sendUserMessageToPlayer(
                sessionId,
                commandHandler,
                'Invalid use of give command. Use "help" for more information.',
            );
            return request.dialogId;
        }

        const result = GiveUserPresetSptCommand.commandRegex.exec(request.text);

        if (!result[1]) {
            this.mailSendService.sendUserMessageToPlayer(
                sessionId,
                commandHandler,
                `Invalid use of give command. Use "help" for more information.`,
            );
            return request.dialogId;
        }

        const profile: ISptProfile = this.saveServer.getProfiles()[sessionId];
        const weaponBuilds = profile.userbuilds.weaponBuilds;
        const weaponBuild = weaponBuilds.find((wb) => wb.Id === result[1]);
        if (!weaponBuild) {
            this.mailSendService.sendUserMessageToPlayer(
                sessionId,
                commandHandler,
                `Couldn't find weapon build for Id: ${result[1]}`,
            );
            return request.dialogId;
        }

        let itemsToSend = this.cloner.clone(weaponBuild.Items);
        itemsToSend = this.itemHelper.replaceIDs(itemsToSend);
        this.itemHelper.setFoundInRaid(itemsToSend);
        this.mailSendService.sendSystemMessageToPlayer(sessionId, "SPT GIVE", itemsToSend);

        return request.dialogId;
    }
}

File: server-mod/GiveUI/Command/GiveUserPresetSptCommand.cs
using System.Text.RegularExpressions;
using SPTarkov.DI.Annotations;
using SPTarkov.Server.Core.Extensions;
using SPTarkov.Server.Core.Helpers;
using SPTarkov.Server.Core.Helpers.Dialog.Commando.SptCommands;
using SPTarkov.Server.Core.Models.Eft.Dialog;
using SPTarkov.Server.Core.Models.Eft.Profile;
using SPTarkov.Server.Core.Servers;
using SPTarkov.Server.Core.Services;
using SPTarkov.Server.Core.Utils.Cloners;

namespace GiveUI.Command;

[Injectable]
public class GiveUserPresetSptCommand(
    MailSendService mailSendService,
    SaveServer saveServer,
    ICloner cloner,
    ItemHelper itemHelper) : ISptCommand
{
    private static readonly Regex _commandRegex = new(@"^spt give-user-preset ((([a-z]{2,5}) )?""(.+)""|\w+)$");

    public string GetCommand()
    {
        return "give-user-preset";
    }

    public string GetCommandHelp()
    {
        return
            "spt give-user-preset\n========\nSends items to the player through the message system.\n\n\tspt give-user-preset [weaponBuilds.Id]";
    }

    public ValueTask<string> PerformAction(UserDialogInfo commandHandler, string sessionId, SendMessageRequest request)
    {
        if (!_commandRegex.IsMatch(request.Text))
        {
            mailSendService.SendUserMessageToPlayer(
                sessionId,
                commandHandler,
                "Invalid use of give command. Use 'help' for more information."
            );
            return new ValueTask<string>(request.DialogId);
        }

        var result = _commandRegex.Match(request.Text);
        var userPresetId = result.Groups[1].Value;
        if (string.IsNullOrEmpty(userPresetId))
        {
            mailSendService.SendUserMessageToPlayer(
                sessionId,
                commandHandler,
                "Invalid use of give command. Use 'help' for more information."
            );
            return new ValueTask<string>(request.DialogId);
        }

        var profile = saveServer.GetProfiles()[sessionId];
        var weaponBuilds = profile.UserBuildData?.WeaponBuilds ?? [];
        var weaponBuild = weaponBuilds.Find((wb) => wb.Id == userPresetId);
        if (weaponBuild == null)
        {
            mailSendService.SendUserMessageToPlayer(
                sessionId,
                commandHandler,
                $"Couldn't find weapon build for Id: {userPresetId}"
            );
            return new ValueTask<string>(request.DialogId);
        }

        var itemsToSend = cloner.Clone(weaponBuild.Items) ?? [];
        itemsToSend = itemsToSend.ReplaceIDs().ToList();
        itemHelper.SetFoundInRaid(itemsToSend);
        mailSendService.SendSystemMessageToPlayer(sessionId, "SPT GIVE", itemsToSend);

        return new ValueTask<string>(request.DialogId);
    }
}